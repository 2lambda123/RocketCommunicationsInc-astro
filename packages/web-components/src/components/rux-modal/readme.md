# rux-modal

A Rux Modal interrupts the app experience to prompt a user to confirm an action or acknowledge a piece of information. It displays information along with a set of buttons allowing users to “Accept or Cancel” the actions presented within the Rux Modal.

## Guidelines

-   [Astro UXDS: Rux Modal](https://astrouxds.com/ui-components/dialog-box)

## Web Components Usage

### 1. Installation

#### Download the [Astro UXDS Component Library](https://github.com/RocketCommunicationsInc/astro-components) source to your project.

Via CLI:

```sh
git clone https://github.com/RocketCommunicationsInc/astro-components.git
```

Or, [download the Astro UXDS Components as a .zip](https://github.com/RocketCommunicationsInc/astro-components/archive/master.zip)

### 2. Import the Astro Rux Modal Web Component

This example assumes you're using the NPM package in `node_modules`. Otherwise, import the component using the path to the Astro Components directory in your project.

```javascript
import { RuxModal } from '@astrouxds/rux-modal/rux-modal.js'
```

### 3. Render the Astro Rux Modal Web Component

At a minimum, Astro UXDS Rux Modal requires a message. In this instance, a single button labeled "Cancel" will be automatically rendered.

```xml
<rux-modal message="This is a message"></rux-modal>
```

Pass properties as attributes of the Astro Rux Modal custom element:

```xml
<rux-modal
  title="Modal title"
  message="Modal message"
  confirmText="Ok"
  denyText="Cancel"
  customEvent="listen-for-me">
</rux-modal>
```

<!-- Auto Generated Below -->


## Properties

| Property       | Attribute        | Description                                 | Type      | Default |
| -------------- | ---------------- | ------------------------------------------- | --------- | ------- |
| `clickToClose` | `click-to-close` | Allows modal to close by clicking off of it | `boolean` | `false` |
| `open`         | `open`           | Shows and hides modal                       | `boolean` | `false` |


## Events

| Event            | Description                           | Type                   |
| ---------------- | ------------------------------------- | ---------------------- |
| `ruxmodalclosed` | Event that is fired when modal closes | `CustomEvent<boolean>` |


## Shadow Parts

| Part          | Description               |
| ------------- | ------------------------- |
| `"(default)"` | the content of the modal  |
| `"footer"`    | the footer of the modal   |
| `"header"`    | the header of the modal   |
| `"wrapper"`   | the modal wrapper overlay |


## CSS Custom Properties

| Name                       | Description            |
| -------------------------- | ---------------------- |
| `--modal-background-color` | Modal background color |
| `--modal-border-color`     | Modal border color     |
| `--modal-title-color`      | Modal title color      |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
