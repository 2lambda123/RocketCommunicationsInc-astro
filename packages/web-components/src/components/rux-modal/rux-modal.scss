@use '../../global/mixins';

:host {
    /**
    * @part wrapper - css part hook to style modal wrapper
    */

    /**
    * @prop --modal-title-color: Modal title color
    */
    --modal-title-color: var(--color-text-primary);

    /**
    * @prop --modal-background-color: Modal background color
    */
    --modal-background-color: var(--color-background-surface-default);

    /**
    * @prop --modal-border-color: Modal border color
    */
    --modal-border-color: var(--color-interactive-default);

    display: contents;
}

*,
*::before,
*::after {
    box-sizing: inherit;
}

//  this styling is only used in storybook up for presentational purposes
//  default flex position takes up the entire body element en spills overlay
//  on top of the docs page

:host([hidden]) {
    display: none;
}

.invisible {
    visibility: hidden;
}

.rux-modal {
    &__wrapper {
        position: fixed;
        top: 0;
        left: 0;
        height: 100vh;
        width: 100vw;
        display: flex;
        justify-content: center;
        align-items: center;
        z-index: 1100;
        background-color: rgba(0, 0, 0, 0.5);
        @include mixins.animation(fadeIn, 0.5s);
    }
    &__dialog {
        position: relative;
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        background-color: var(--modal-background-color);
        width: 28rem;
        min-height: 13.5rem;
        border: 0;
        margin: auto;
        padding: 0;
        user-select: none;
        box-shadow: var(--shadow-outer-dialog);
    }

    &__titlebar {
        display: flex;
        flex-grow: 0;
        flex-shrink: 0;
        align-items: center;
        width: 100%;
        padding: 16px 20px;
        background-color: var(--color-background-surface-header);
        color: var(--modal-title-color);
        user-select: none;
        font-family: var(--font-heading-2-font-family);
        font-size: var(--font-heading-2-font-size);
        font-weight: var(--font-heading-2-font-weight);
        letter-spacing: var(--font-heading-2-letter-spacing);
        // Allows h tags to have same amount of margin as any other tag
        ::slotted(h6),
        ::slotted(h5),
        ::slotted(h4),
        ::slotted(h3),
        ::slotted(h2),
        ::slotted(h1) {
            margin-block-start: 0;
            margin-block-end: 0;
        }
    }

    &__content {
        padding: 0 1.25rem; //0 20px
        color: var(--color-default-text);
    }

    &__footer {
        padding: 1.25rem; //20px
    }
}

rux-icon {
    margin-right: 0.75rem;
}

@include mixins.keyframes(fadeIn) {
    0% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
}
