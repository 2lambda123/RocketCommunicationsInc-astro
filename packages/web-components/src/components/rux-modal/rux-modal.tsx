import {
    Component,
    h,
    Prop,
    Event,
    EventEmitter,
    Element,
    Listen,
    Watch,
    Host,
} from '@stencil/core'
import { hasSlot } from '../../utils/utils'

/**
 * @part wrapper - the modal wrapper overlay ! DEPRECATED IN FAVOR OF CONTAINER !
 * @part container - the modal container
 * @part dialog - the native dialog element
 * @part header - the header of the modal
 * @part message - the message of the modal
 * @part confirm-button - the modal's confirm button
 * @part deny-button - the modal's deny button
 * @part footer - the footer of the modal
 *
 * @slot header - the header of the modal
 * @slot message - the modal's message or content
 * @slot footer - the footer of the modal
 */
@Component({
    tag: 'rux-modal',
    styleUrl: 'rux-modal.scss',
    shadow: true,
})
export class RuxModal {
    /**
     * Shows and hides modal
     */
    @Prop({ reflect: true, mutable: true }) open: boolean = false
    /**
     * Modal body message
     */
    @Prop() modalMessage?: string
    /**
     * Modal header title
     */
    @Prop() modalTitle?: string
    /**
     * Text for confirmation button
     */
    @Prop() confirmText: string = 'Confirm'
    /**
     * Text for close button
     */
    @Prop() denyText: string = 'Cancel'
    /**
     * Event that is fired when modal opens
     */
    @Event({
        eventName: 'ruxmodalopened',
        composed: true,
        bubbles: true,
    })
    ruxModalOpened!: EventEmitter<void>
    /**
     * Event that is fired when modal closes
     */
    @Event({
        eventName: 'ruxmodalclosed',
        composed: true,
        bubbles: true,
    })
    ruxModalClosed!: EventEmitter<boolean>

    @Element() element!: HTMLRuxModalElement

    private hasFooter = hasSlot(this.element, 'footer')
    private hasHeader = hasSlot(this.element, 'header')
    private hasMessage = hasSlot(this.element, 'message')

    // confirm dialog if Enter key is pressed
    @Listen('keydown', { target: 'window' })
    handleKeyDown(ev: KeyboardEvent) {
        if (ev.key === 'Enter') {
            const button = this._getDefaultButton()
            if (button) {
                button.click()
            }
        }
    }

    // close modal if click happens outside of dialog
    @Listen('click', { target: 'window' })
    handleClick(ev: MouseEvent) {
        const wrapper = this._getWrapper()
        if (ev.composedPath()[0] === wrapper) {
            // this.ruxModalClosed.emit(false)
            this.open = false
        }
    }

    @Watch('open')
    handleOpen(isOpen: boolean) {
        //? This is preventing the default focusing of a button if the footer slot exists.
        //? I thought about runninng it anyway, but it could lead to cases where an unintentional button gets focused.
        if (isOpen && !this.hasFooter) {
            setTimeout(() => {
                const button = this._getDefaultButton()
                if (button) {
                    button.focus()
                }
            }, 0)
        }
        this.open ? this.ruxModalOpened.emit() : this.ruxModalClosed.emit()
    }

    private _handleModalChoice() {
        this.open = false
    }

    private _getDefaultButton(): HTMLElement | null {
        const buttonSet = this.element?.shadowRoot?.querySelectorAll(
            'rux-button:not([hidden])'
        ) as NodeListOf<HTMLElement>

        if (buttonSet.length > 0) {
            const defaultButton = buttonSet[buttonSet.length - 1]
            //* Need to get the native <button> to focus. This isn't necessary if we add the delegateFocus option to the stencil shadow in Stencil 2.10
            const shadow = defaultButton.shadowRoot?.querySelector('button')

            if (shadow) return shadow
        }

        return null
    }

    private _getWrapper(): HTMLElement | null {
        const wrapper = this.element?.shadowRoot?.querySelector(
            '.rux-modal__wrapper'
        ) as HTMLElement

        if (wrapper) {
            return wrapper
        }
        return null
    }

    connectedCallback() {
        this._handleModalChoice = this._handleModalChoice.bind(this)
    }

    componentDidRender() {
        const button = this._getDefaultButton()
        button && button.focus()
    }

    render() {
        const {
            open,
            modalMessage,
            modalTitle,
            confirmText,
            denyText,
            _handleModalChoice,
        } = this

        return (
            open && (
                <Host>
                    <div part="wrapper container" class="rux-modal__wrapper">
                        <dialog
                            class="rux-modal__dialog"
                            role="dialog"
                            part="dialog"
                        >
                            {modalTitle || this.hasHeader ? (
                                <header class="rux-modal__header" part="header">
                                    <slot name="header">{modalTitle}</slot>
                                </header>
                            ) : null}
                            {modalMessage || this.hasMessage ? (
                                <div class="rux-modal__content" part="message">
                                    <div
                                        class="rux-modal__message"
                                        part="message"
                                    >
                                        <slot name="message">
                                            {modalMessage}
                                        </slot>
                                    </div>
                                </div>
                            ) : null}
                            {this.hasFooter ? (
                                <footer class="rux-modal__footer" part="footer">
                                    <slot name="footer"></slot>
                                </footer>
                            ) : (
                                <footer class="rux-modal__footer" part="footer">
                                    <rux-button-group h-align="right">
                                        <rux-button
                                            secondary={confirmText.length > 0}
                                            onClick={_handleModalChoice}
                                            hidden={!denyText}
                                            tabindex="-1"
                                            exportparts="container:deny-button"
                                        >
                                            {denyText}
                                        </rux-button>
                                        <rux-button
                                            onClick={_handleModalChoice}
                                            hidden={!confirmText}
                                            tabindex="0"
                                            exportparts="container:confirm-button"
                                        >
                                            {confirmText}
                                        </rux-button>
                                    </rux-button-group>
                                </footer>
                            )}
                        </dialog>
                    </div>
                </Host>
            )
        )
    }
}
