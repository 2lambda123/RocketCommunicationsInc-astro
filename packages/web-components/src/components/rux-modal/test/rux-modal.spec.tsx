import { h } from '@stencil/core'
import { newSpecPage, SpecPage } from '@stencil/core/testing'
import { RuxModal } from '../rux-modal'

describe('rux-modal', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxModal],
            html: `<rux-modal></rux-modal>`,
        })
        expect(page.root).toEqualHtml(`
            <rux-modal>
                <mock:shadow-root></mock:shadow-root>
            </rux-modal>
        `)
    })
    it('should emit one event when closed', async () => {
        const clickSpy = jest.fn()
        let page: SpecPage

        page = await newSpecPage({
            components: [RuxModal],
            template: () => (
                <div>
                    <div id="click">Something to click</div>
                    <rux-modal
                        click-to-close
                        open
                        onRuxmodalclosed={(ev: any) => clickSpy(ev)}
                    >
                        <div slot="header">Header</div>
                    </rux-modal>
                </div>
            ),
        })!

        page.doc?.addEventListener('onruxclosed', () => {
            console.log('heard')
        })
        const modal = page.doc?.querySelector('rux-modal')
        expect(modal!.open).toBe(true)
        const div = page.doc?.getElementById('click')
        // modal!.open = false
        div!.click()
        // expect(modal!.open).toBe(false)
        expect(clickSpy).toBeCalledTimes(1)
    })
})
