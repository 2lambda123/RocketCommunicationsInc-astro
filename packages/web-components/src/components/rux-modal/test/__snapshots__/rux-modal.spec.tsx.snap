// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-modal renders with a mix of slots and props 1`] = `
<rux-modal modal-title="Title">
  <mock:shadow-root></mock:shadow-root>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-modal>
`;

exports[`rux-modal renders with a mix of slots and props 2`] = `
<rux-modal modal-title="Title">
  <mock:shadow-root></mock:shadow-root>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-modal>
`;

exports[`rux-modal renders with a mix of slots and props 3`] = `
<rux-modal modal-title="Title">
  <mock:shadow-root></mock:shadow-root>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-modal>
`;

exports[`rux-modal renders with slots 1`] = `
<rux-modal>
  <mock:shadow-root></mock:shadow-root>
  <span slot="header">
    Header
  </span>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-modal>
`;

exports[`rux-modal renders with slots 2`] = `
<rux-modal>
  <mock:shadow-root></mock:shadow-root>
  <span slot="header">
    Header
  </span>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-modal>
`;

exports[`rux-modal renders with slots 3`] = `
<rux-modal>
  <mock:shadow-root></mock:shadow-root>
  <span slot="header">
    Header
  </span>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-modal>
`;
